### PR TITLE
Add mlperf RNN-T model

### PR DESCRIPTION
**Pull request by wozeparrot**
For now dataset has to be downloaded manually from https://www.openslr.org/12/. mlperf uses `dev-clean`.
Then all the flacs have to be converted to wav using something like:
```fish
for file in **/*.flac; ffmpeg -i $file -ar 16k "$(dirname $file)/$(basename $file .flac).wav"; end
```
Need this file as well https://github.com/mlcommons/inference/blob/master/speech_recognition/rnnt/dev-clean-wav.json.

I'm currently using `librosa` for the stft and audio file loading but it might be possible to rewrite them to not rely on librosa.
The RNN-T bench is also slightly different from the other ones as feature extraction is part of the benchmark, so I'm not quite sure where to put the feature extraction function.

Edit: this is also pretty slow rn
Edit again: not slow anymore

### CODE REVIEW

**Comment by radenmuaz** [comment]
was writing rnnt till I see this commit...
btw how is the WER on dev set?

File: models/rnnt.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  def __call__(self, x, h, c):
+    t = (x @ self.weights_ih.T) + self.bias_ih + (h @ self.weights_hh.T) + self.bias_hh
+
+    i, f, g, o = t.chunk(4, 1)
+    i, f, g, o = i.sigmoid(), f.sigmoid(), g.tanh(), o.sigmoid()
+
+    c = (f * c) + (i * g)
+    h = o * c.tanh()
+
+    return h.realize(), c.realize()
```
Comment (geohot) [inline models/rnnt.py:102]: Do you have to realize here?

File: models/rnnt.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  def __init__(self, input_size, hidden_size, layers, dropout):
+    self.input_size = input_size
+    self.hidden_size = hidden_size
+    self.layers = layers
+    self.dropout = dropout
+
+    self.cells = [LSTMCell(input_size, hidden_size) if i == 0 else LSTMCell(hidden_size, hidden_size) for i in range(layers)]
+
+  def __call__(self, x, h, c):
+    if h is None or c is None:
```
Comment (geohot) [inline models/rnnt.py:115]: Creating tensors is slow.

Why do you have to make copies?

x.assign(<new tensor>) is better than x = <new tensor>

File: models/rnnt.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    return labels
+
+  def load_from_pretrained(self):
+    fn = Path(__file__).parent.parent / "weights/rnnt.pt"
+    download_file("https://zenodo.org/record/3662521/files/DistributedDataParallel_1576581068.9962234-epoch-100.pt?download=1", fn)
+
+    import torch
+    state_dict = torch.load(open(fn, "rb"), map_location="cpu")["state_dict"]
+
+    # encoder
```
Comment (geohot) [inline models/rnnt.py:57]: See how the other models do this with get_child. Any reason this one can't be done like that? The code is a lot shorter

Comment (wozeparrot) [reply models/rnnt.py:57]: torch uses `weights_ih_l{layer_number}` for their lstm layer weights which can't really be done cleanly

File: models/rnnt.py (hunk 2 above)
Comment (wozeparrot) [reply models/rnnt.py:115]: the copy here is actually a copy of the list not a copy of the tensors

Comment (wozeparrot) [reply models/rnnt.py:115]: also we need the zero tensors here are this is state that is passed between the layers

File: models/rnnt.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+
+class LSTMCell:
+  def __init__(self, input_size, hidden_size):
+    self.hidden_size = hidden_size
+
+    self.weights_ih = Tensor.uniform(hidden_size * 4, input_size)
+    self.bias_ih = Tensor.uniform(hidden_size * 4)
+    self.weights_hh = Tensor.uniform(hidden_size * 4, hidden_size)
+    self.bias_hh = Tensor.uniform(hidden_size * 4)
```
Comment (ym1234) [inline models/rnnt.py:91]: (Not a review just a question) Why use tensors instead of nn.Linear here?

Comment (wozeparrot) [reply models/rnnt.py:91]: maybe its cleaner to use linear here but the weight initialization for linear is different than the weight initialization for LSTM at least in pytorch.

**Comment by geohot** [comment]
What's the current val loss?

**Comment by wozeparrot** [comment]
Currently doing another run cuz I found that the metric code was wrong

**Comment by wozeparrot** [comment]
ok, I got a word error rate of 7.38024337340539% across the filtered dev-clean set that mlperf specifies. the mlperf word error rate is specified as 7.452253714852645%.

File: models/rnnt.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        h = inp[BS*self.input_size + i*BS*self.hidden_size:BS*self.input_size + (i+1)*BS*self.hidden_size].reshape(BS, self.hidden_size)
+        c = inp[BS*self.input_size + BS*self.hidden_size*self.layers + i*BS*self.hidden_size:BS*self.input_size + BS*self.hidden_size*self.layers + (i+1)*BS*self.hidden_size].reshape(BS, self.hidden_size)
+        x, h_, c_ = cell(x, h, c)
+        new_h.append(h_)
+        new_c.append(c_)
+      return x.cat(*new_h, *new_c).realize()
+
+  return _LSTM
+
+LSTM = _FastLSTM()
```
Comment (geohot) [inline models/rnnt.py:143]: Is this extra layer of indirection needed?

File: datasets/librispeech.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  return features.transpose(2, 0, 1), x_lens.astype(np.float32)
+
+def load_wav(file):
+  sample = librosa.load(file, sr=16000)[0].astype(np.float32)
+  return sample, sample.shape[0]
+
+def iterate(bs=8, start=0):
+  print(f"there are {len(ci)} samples in the dataset")
+  for i in range(start, len(ci), bs):
+    samples, sample_lens = zip(*[load_wav(os.path.join(BASEDIR, v["files"][0]["fname"])) for v in ci[i : i + bs]])
```
Comment (geohot) [inline datasets/librispeech.py:56]: Should probably be in a subdirectory

**Comment by geohot** [comment]
Great work.

I think the last thing is a dataset downloader. And it's a nitpick, but in general, use pathlib instead of os.path.join.

I'll test it, if I repro the WER, the $200 bounty is yours. How do you want to be paid? (can also claim privately via e-mail)

**Comment by geohot** [comment]
Oh, test is failing.

**Comment by wozeparrot** [comment]
> Oh, test is failing in LLVM + GPU.

GPU test failure doesn't look related, will take a look at llvm

**Comment by geohot** [comment]
The GPU failure is likely related. It's because RNN-T isn't being cleaned up cleanly, there's tensors stuck somewhere.

**Comment by wozeparrot** [comment]
Both tests should be fixed now

File: datasets/librispeech.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import json
+import pathlib
+import numpy as np
+import librosa
+
+BASEDIR = pathlib.Path("/home/woze/projects/tinygrad/datasets/librispeech")
```
Comment (geohot) [inline datasets/librispeech.py:6]: Use __file__ to not code directories specific to your machine.

Also, I'm fine with not having an auto download script for the dataset, but include your top GitHub comment as a comment in here so it works standalone.

**Comment by geohot** [comment]
Tests also still failing.

Ugh, I wish I could disable "Approve and Run" on Github. https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/

**Comment by geohot** [comment]
Testing now, just my notes:
```
for file in $(find * | grep flac); do ffmpeg -i $file -ar 16k "$(dirname $file)/$(basename $file .flac).wav"; done
ln -s LibriSpeech/dev-clean/ dev-clean-wav
```

Model download is super slow...

**Comment by geohot** [comment]
Nice work, seeing WER around 0.04, bounty is yours!

Let me know how you want to be paid (george@tinygrad.org)
